### PR TITLE
Implement full-screen editor modal and address review feedback

### PR DESCRIPTION
**Pull request by OPS-PIvers**

### CODE REVIEW

File: components/widgets/Schedule/ScheduleWidget.tsx [hunk 1]
```diff
@@ -317,26 +336,49 @@ export const ScheduleWidget: React.FC<{ widget: WidgetData }> = ({
...
   // re-launch on every interval tick.
   const launchedItemsRef = useRef<Set<string>>(new Set());
 
+  // Keep the latest effective times accessible to the auto-launch interval
+  // (which runs on a 10s tick but wants the current chain).
+  const effectiveTimesRef = useRef(effectiveTimes);
+  useEffect(() => {
+    effectiveTimesRef.current = effectiveTimes;
+  }, [effectiveTimes]);
+
```
Comment (OPS-PIvers) [inline components/widgets/Schedule/ScheduleWidget.tsx:373]: Using `useEffect` to sync `effectiveTimesRef.current` introduces an unnecessary render-then-effect delay before the interval sees the updated chain, and violates the pattern documented in CLAUDE.md:

> Assign refs directly in the render body: `myRef.current = value` — no effect needed.

Drop the `useEffect` and assign directly after the `useRef` declaration:

```ts
const effectiveTimesRef = useRef(effectiveTimes);
effectiveTimesRef.current = effectiveTimes;  // always in sync, no effect needed
```

The same issue exists for `configRef`, `itemsRef`, `scheduleItemsRef`, and `widgetRef` in this file (pre-existing, but this PR shouldn't add a new instance of the pattern). All five of those refs are only read inside interval/timeout callbacks that already fire after the render, so the one-frame lag from `useEffect` is harmless in practice — but the render-body assignment is both correct and simpler.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by OPS-PIvers — COMMENTED** [review]
One new issue in `ScheduleWidget.tsx` not covered by the existing two open review threads.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: components/widgets/Embed/Widget.tsx [hunk 2]
```diff
@@ -81,6 +85,100 @@ export const EmbedWidget: React.FC<{ widget: WidgetData }> = ({ widget }) => {
...
+
+    const mutationObserver = new MutationObserver(() => updateRect());
+    mutationObserver.observe(widgetEl, {
+      attributes: true,
+      attributeFilter: ['style'],
+    });
+
+    const onScrollOrResize = () => updateRect();
+    window.addEventListener('scroll', onScrollOrResize, true);
+    window.addEventListener('resize', onScrollOrResize);
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply components/widgets/Embed/Widget.tsx]: Not actionable — the scroll/resize handler on `EmbedWidget` is already throttled to animation-frame cadence via `requestAnimationFrame`. Every event (scroll, resize, ResizeObserver, MutationObserver) feeds into a single `scheduleUpdate` that no-ops if a rAF is already pending, so `getBoundingClientRect()`/`setState` fires at most once per frame — tighter than a typical `use-debounce` interval and the standard "match the display refresh" pattern for positional sync. Closing as already addressed.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: components/common/EditorModalShell.tsx [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    isSaving,
+    onClose,
+    showConfirm,
+    confirmDiscardMessage,
+    confirmDiscardTitle,
+  ]);
+
+  const handleSave = useCallback(async () => {
+    if (saveDisabled || isSaving) return;
+    await onSave();
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply components/common/EditorModalShell.tsx]: Already addressed — `handleSave` already wraps `await onSave()` in try/catch (see `components/common/EditorModalShell.tsx`). On rejection we log to `console.error` and surface a toast via the existing `saveErrorMessage` prop (callers can set it to `false` to suppress the toast when they render their own inline error). The `void handleSave()` click handler therefore cannot produce an unhandled promise rejection. Closing as complete.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: components/widgets/Schedule/ScheduleWidget.tsx (hunk 1 above)
Comment (OPS-PIvers) [reply components/widgets/Schedule/ScheduleWidget.tsx:373]: Declining — although CLAUDE.md documents `myRef.current = value` in the render body as the preferred pattern, this repo's ESLint config enables `react-hooks/refs`, which actively errors on render-body ref assignment:

```
371:3  error  Cannot update ref during render  react-hooks/refs
```

I tried the render-body fix locally and it fails `pnpm run lint` (and would block CI via the `--max-warnings 0` gate). The existing `useEffect` synchronization satisfies the lint rule, and — as the original comment notes — the one-frame lag is harmless because `effectiveTimesRef.current` is only read inside the 10 s auto-launch interval, not during render.

Leaving the pattern as-is; CLAUDE.md should probably be updated to reflect that the render-body rule is opt-in (and currently not in effect due to this lint rule), but that's out of scope for this PR.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by OPS-PIvers — COMMENTED** [review]
## Automated Code Review

**Date:** 2026-04-16

### Summary
Large PR (51 files, +4766/-1588) implementing a full-screen editor modal pattern, Quiz assignment system, Schedule widget improvements, Embed widget portal-based toolbar, Dice color customization, and Video Activity editor migration. The scope is ambitious — several distinct features in one PR. Code quality is generally high, but there are a few concerns worth addressing.

### Code Quality
- **DiceWidget regression — hardcoded `fontSize: '20px'`** on the Roll button: This is a `skipScaling: true` widget, so front-face content should use container query units per project standards. The previous `min(20px, 5cqmin)` was correct; the change to fixed `20px` means the button text won't scale down in small widget sizes.
- **EmbedWidget `console.warn`** (`[EmbedWidget] No [data-widget-id=...]`): Not gated behind `import.meta.env.DEV`. Should be removed or dev-gated to avoid production console noise.
- **QuizAssignmentSettingsModal `console.error`** on save failure: Already surfaced via toast — the `console.error` is redundant in production. Consider gating behind `import.meta.env.DEV`.
- **VideoActivityWidget type cast** (`config.view as VideoActivityView | 'editor' | undefined`): Used as a migration bridge for the removed `'editor'` view value. Functionally correct but suppresses type safety. A runtime check (`if (view === 'editor') view = 'manager'`) would be cleaner.
- **`useQuizAssignments` snapshot handler**: Uses `d.data() as QuizAssignment` without spreading `d.id`. If any document is created without embedding `id` in the payload (e.g., via Cloud Function or console), the `id` field will be `undefined`. Consider `{ ...d.data(), id: d.id } as QuizAssignment` for robustness.

### Widget Standards
- No new widget registrations needed (existing widgets modified, not added).
- Schedule widget extracted utilities to `Schedule/utils.ts` with thorough test coverage — good pattern.
- EmbedWidget toolbar is now portaled outside the container-query scope, so fixed Tailwind classes (`w-4 h-4`) are acceptable there.
- DiceWidget Roll button font size regression noted above.

### Admin Config Standards
Not applicable — no admin configuration panels were added or modified.

### Firestore & Security
- **`quiz_assignments` collection (new, user-scoped):** Well secured. Enforces `request.auth.uid == userId`. Immutability on `teacherUid` and `id` fields on create/update prevents ownership spoofing.
- **`shared_assignments` collection (new, global):** Follows existing `shared_quizzes` pattern. Only original author can create; only author or admin can update/delete. `list` is denied by default.
- **`quiz_sessions` refactored to random `sessionId`:** Ownership now checked via `resource.data.teacherUid` field (not path). Correctly enforced on create, update, and delete. The `sessionTeacherUid()` helper uses `get()` calls for response sub-rules — counts against the 10-call-per-rule Firestore limit. Not an issue today but could be if more nested rules are added.
- **No auth bypasses or overly permissive rules.**
- **Composite index requirement:** The `allocateJoinCode()` query in `useQuizAssignments.ts` uses `where('code', '==', ...), where('status', '!=', 'ended')` — requires a composite Firestore index. Verify this exists in Firestore indexes config or console.

### Regression Risk
- **HIGH — `QuizSession.id` semantics changed** from teacher UID to session UUID. Any code relying on `session.id === user.uid` for Firestore lookups will silently break. Verify all consumers of `QuizSession.id` across the codebase.
- **MEDIUM — `VideoActivityView` removes `'editor'` value.** The migration cast handles this, but confirm no remaining references to `'editor'` in video activity code paths.
- **LOW — DashboardContext changes** are additive only (new `pendingAssignmentShareId` state). Existing share URL handling preserved.
- **LOW — WidgetRegistry.ts not modified.** No registry consistency risk.

### Missing Tests
Significant test coverage gap for new code:

| File | New Lines | Tests? |
|------|-----------|--------|
| `hooks/useQuizAssignments.ts` | +462 | None |
| `QuizAssignmentArchive.tsx` | +472 | None |
| `QuizAssignmentSettingsModal.tsx` | +497 | None |
| `utils/imageProcessing.ts` | +71 | None |

~1,500 lines of new production code have no test coverage. The `useQuizAssignments` hook contains complex Firestore CRUD and real-time listener logic that would benefit most from unit tests. `imageProcessing.ts` is a pure utility that would be trivial to test.

Positive: `Schedule/utils.test.ts` (+209 lines) is thorough with good edge case coverage. Existing test files received appropriate updates.

### Overall
**Merge readiness:** Needs changes

The PR delivers valuable features but has three items that should be addressed before merge:
1. **Verify `QuizSession.id` semantics change** doesn't break any existing consumers (highest risk)
2. **Fix DiceWidget Roll button scaling regression** (`fontSize: '20px'` → `min(20px, 5cqmin)`)
3. **Confirm composite Firestore index** exists for `allocateJoinCode()` query

The test coverage gap for the Quiz Assignment system (~1,400 lines) is notable but may be acceptable depending on team testing strategy.

---
_This review was generated automatically. All findings should be verified by a human before merging._

---
_Generated by [Claude Code](https://claude.ai/code)_
